### PR TITLE
sign: throw on unsupported padding scheme

### PR DESCRIPTION
**Pull request by btj**
Do not silently apply the wrong padding scheme. Would have saved me hours of head-scratching and screen-staring.

### CODE REVIEW

File: browser/sign.js [hunk 1]
```diff
@@ -19,6 +21,7 @@ function sign (hash, key, hashType, signType, tag) {
   } else {
     if (signType !== 'rsa' && signType !== 'ecdsa/rsa') throw new Error('wrong private key type')
   }
+  if (key.padding !== undefined && key.padding !== RSA_PKCS1_PADDING) throw new Error('padding scheme not supported')
```
Comment (ljharb) [inline browser/sign.js:24]: can you add some tests for this? What does node do in this case?

**Review by ljharb — CHANGES_REQUESTED** [review]

File: browser/sign.js (hunk 1 above)
Comment (btj) [reply browser/sign.js:24]: Node.js also [supports](https://nodejs.org/api/crypto.html#signsignprivatekey-outputencoding) `RSA_PKCS1_PSS_PADDING`. If the padding value is incorrect, the [call](https://github.com/nodejs/node/blob/18e00a577d7405ca0e6c26113675c230677342c2/src/crypto/crypto_sig.cc#L64) of OpenSSL's [`EVP_PKEY_CTX_set_rsa_padding`](https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_CTX_set_rsa_pss_saltlen.html) returns a negative value and Node.js presumably throws an exception.

Comment (ljharb) [reply browser/sign.js:24]: Great - then i should be able to verify that with the test case for this package :-)

File: test/index.js [hunk 2]
```diff
@@ -48,6 +48,39 @@ fixtures.valid.rsa.forEach(function (f) {
   });
 });
 
+(function () {
```
Comment (ljharb) [inline test/index.js:51]:
```suggestion
test('padding schemes', function (t) {
```

File: test/index.js [hunk 3]
```diff
@@ -48,6 +48,39 @@ fixtures.valid.rsa.forEach(function (f) {
...
+  var f = fixtures.valid.rsa[0];
+  var message = Buffer.from(f.message);
+  var priv = {
+    key: Buffer.from(f['private'], 'base64'),
+    padding: 11646841 // Some invalid value
+  };
+
+  test('invalid padding option', function (t) {
```
Comment (ljharb) [inline test/index.js:59]:
```suggestion
  t.test('invalid padding option', function (st) {
```

File: test/index.js [hunk 4]
```diff
@@ -48,6 +48,39 @@ fixtures.valid.rsa.forEach(function (f) {
...
+    }
+
+    t.ok(nExc, 'node throws an exception');
+    t.equal(String(nExc), 'Error: error:1C8000A5:Provider routines::illegal or unsupported padding mode', 'node throws exception with proper message');
+    t.ok(bExc, 'browser throws an exception');
+    t.equal(String(bExc), 'Error: illegal or unsupported padding mode', 'browser throws exception with proper message');
+
+    t.end();
+  });
+}());
```
Comment (ljharb) [inline test/index.js:82]:
```suggestion
    st.ok(nExc, 'node throws an exception');
    st.equal(String(nExc), 'Error: error:1C8000A5:Provider routines::illegal or unsupported padding mode', 'node throws exception with proper message');
    st.ok(bExc, 'browser throws an exception');
    st.equal(String(bExc), 'Error: illegal or unsupported padding mode', 'browser throws exception with proper message');

    st.end();
  });
  
  t.end();
});
```

**Review by ljharb — APPROVED** [review]
awesome, thanks! i'm happy to make the changes i suggested as part of the rebase i'll do once i land another PR that's queued up, and then this can be landed too :-)

**Comment by ljharb** [comment]
yay, the tests are doing their job :-)

there's some failures, where older node isn't throwing an exception where the tests expect it i think?

**Comment by btj** [comment]
Actually, it seems that the check of the error message was too strict; I replaced it by a regex.

**Comment by ljharb** [comment]
a couple more failures?

**Comment by ljharb** [comment]
the 3 remaining failures seem legitimate

**Comment by btj** [comment]
First of all, the test messages were wrong: I had swapped 'browser' and 'node'. Old versions of Node.js do not support the padding option and therefore ignore it and never throw. Programs written against old Node versions will not include the padding option and therefore are not affected by the new throw statement in browserify.

**Comment by btj** [comment]
OK, it turned out the `isNode10()` function in the test suite is broken: it looks at the *minor* version instead of the major version. I fixed it so that the test suite now considers the major version.

Interestingly, this means the passphrase tests *did* run on many of the old Node versions tested, just not on 4.9 and 7.10. And they succeeded! Still, I now disabled these tests for major versions <= 10, as seemed to be intended.

**Comment by btj** [comment]
I now ran the passphrase tests on 4.9 on my machine, and they run just fine, so it seems better to simply enable them always. People might be depending on them working on old node versions.

**Comment by ljharb** [comment]
Good catch! Let's switch those checks to use semver (v6), rather than string munging.

**Comment by btj** [comment]
Weird; I'm unable to reproduce the failure on node 8.0. On my machine, the test suite runs fine on node 8.0.

**Comment by ljharb** [comment]
It works on node 8 latest, but fails for me on node 8.2, for example.

Support for the padding option was indeed added in 8.0, though, so maybe it was broken until v8.6? i'll update the skip logic.

**Pull request merged**